### DISCUSSION TITLE
Discarding artifacts feature: need buttons

### DISCUSSION THREAD

**Discussion by chrisbigart** [ihhub/fheroes2#3674] [Q&A]
Hi, I'm working on a feature to allow discarding artifacts (with the exception of magic book, fizbin, tax lien, and hideous mask) in the hero screen and after a victory:

![image](https://user-images.githubusercontent.com/17466060/121772879-46bef100-cb46-11eb-8a6d-cc0e8df378b5.png)

![image](https://user-images.githubusercontent.com/17466060/121772948-cea4fb00-cb46-11eb-970c-d78bb1ceaf23.png)

The code is pretty much ready to go, but I need help with the buttons.  I don't know if a button with an icon would be better, or just simply a button with the text 'Discard' on it.  Maybe for the victory condition, there could be 'Discard' and 'Keep' instead of 'Okay' and a discard button.  In any case, if someone could help me create these or provide them, then I'll be able to submit this change.  Thanks!

**Comment by ihhub** [comment]
Hi @chrisbigart , thanks for opening a discussion before proposing an implementation. From what see many people will be against it as it's breaking game's logic. I'm sure some people in our team will share the same opinion so I'll just wait to hear them🙂

**Reply by chrisbigart** [reply]
Right now I have implemented this as an additional 'experimental feature' so players would have to opt-in to it.

The most annoying thing about H2 for me personally, especially on larger maps, is having to micro-manage my main hero's artifacts when the backpack is full or nearly full (buying another hero to transfer lesser artifacts just to dismiss them, transfering artifacts to a temporary hero before a fight so I can capture as many enemy artifacts as possible, etc).  Another solution (proposed by others in the issues) is to have a secondary backpack (similar to H3), but that is an even bigger gameplay change, and more complex to implement so I am starting with this proposal.

**Reply by Branikolog** [reply]
Hello, @chrisbigart !
Really appreciate your desire to improve the game.
But I'm afraid, that feature you requested is too drastic. Limited inventory is a distinctive feature of Heroes of Might and Magic 2. Allowing player to dismiss some artifacts (and inability to not to dismiss other because of exceptions) will severely ruin the game balance and looks like a feature that turns the game into something another... Originally, Heroes 2 game doesn't have so many artifacts, that can cause an excess in hero inventory. This limitation is a part of the game, and implementing tons of features like you've proposed will turn this brilliant game into game-of-1000-options, which is not user-friendly.
Yes. I know, that fheroes2 engine currently has some alike experimental features, but they were implemented during previous development not by our team.  These features don't fit original game style of Heroes 2 and even were implemented in not a proper way and may work not correctly.
Currently, project development is focused on fixing bugs and implementing "quality of life" features, making gameplay comfortable, rather than adding new ones changing rules of the game.

**Reply by LeHerosInconnu** [reply]
Hello everyone,

Although I would prefer a more advanced option for artifact management (for all the reasons mentioned in the beginning of this thread: https://github.com/ihhub/fheroes2/discussions/2995), the option proposed here goes in the right direction.

I've watched a lot of videos about Heroes 2 and many players would like to have this option to get rid of artifacts that clutter their hero's inventory.

While this is part of the design of the original game, it is already possible in the original game to use another hero to get rid of artifacts, but this is in my opinion just a way to remedy a design flaw of the game in a not very interesting way with boring micro-management.
Especially with the extra artifacts in the expansion and all those added spell scrolls, the hero inventory of the original game doesn't really fit anymore.
This has been corrected in Heroes 3 by the way.

That's why I think this option is a good start and could be added to the game.
I'm not sure I'll use it because I'm reluctant to have an artifact disappear from the game (I prefer to find another use for them), but other players won't hesitate for a second.

I would also like to add some suggestions regarding this option:
- I suggest that the button be the same type as the "Hero" button and with an icon rather than text (no need to change it for translations).
An arc of a circle with an arrow at the bottom (to symbolize throwing an object) or a straight arrow pointing down would be enough I think.
![Drop artifact 001](https://user-images.githubusercontent.com/43583994/121778969-b6120080-cb99-11eb-9a7b-b98174cdc851.png)

- The number of free/occupied slots in the hero's inventory should also be clearly indicated above the button, e.g. "3/14" (in the same way as the number of free/used secondary skills above the "Hero" button).
(I would prefer the smaller font and also for the "Hero" button.)
- This possibility to get rid of an artifact should be added in all cases where the hero receives an artifact.
![Drop artifact 002](https://user-images.githubusercontent.com/43583994/121778977-c4601c80-cb99-11eb-9814-ab2d7fc321fa.png)

- Also a "Hero" button should be added to allow to consult the hero's inventory.
- It should also be possible in the hero's screen (no need for the "Hero" button here).
![Drop artifact 003](https://user-images.githubusercontent.com/43583994/121778991-d04bde80-cb99-11eb-8382-1bfc26caf670.png)

- The button should be grayed out when the artifact is needed for the scenario, with a right-click information window indicating why (e.g. victory condition).
- The function of the button is also indicated in an information window with the right click.

@Branikolog.

> ...implementing tons of features like you've proposed will turn this brilliant game into game-of-1000-options, which is not user-friendly.

It's just the opposite. :)
If an option brings something interesting and/or something that is requested, why would you want to limit the number of options.
It makes no sense to me to limit the number of options because there are too many, the interface should present them in a way that is not overwhelming.

> Currently, project development is focused on fixing bugs and implementing "quality of life" features, making gameplay comfortable, rather than adding new ones changing rules of the game.

In a sense, this option can be seen as a "quality of life" improvement. :D

**Edit.**

A button like this one:

![Bouton drop 01](https://user-images.githubusercontent.com/43583994/121780759-16a53b80-cba2-11eb-96b0-30b44cb30cf5.png)

Also a confirmation window with an "OKAY" button and a "CANCEL" button should be displayed.

**Reply by Branikolog** [reply]
Hi,  @LeHerosInconnu !
For now we're working on 99% pure version of Heroes 2.
In future, when engine will be ready, I believe multiple versions of Heroes 2 will appear, including the one, that'd offer a significant number of options, changing rules of the whole game. 
But if we focus on adding such kind of features now - engine will never be ready. ;) 
I'm not opposing to add new features like above. It's just too early.

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

It's better to look at it that way. :)

**Reply by ihhub** [reply]
If we want to add such option then I'd like to ask how AI would behave in this case? I'm the proposed changes cover only human player. Another issue to solve: why negative artifacts can't be removed? It needs a proper explanation from the game. Also we have PoL artifacts which have positive and negative effects. How are we going to handle them?

**Reply by LeHerosInconnu** [reply]
Hello @ihhub,

> If we want to add such option then I'd like to ask how AI would behave in this case? I'm the proposed changes cover only human player.

I think the AI should keep all artifacts (and distribute them intelligently among its heroes); a bad artifact in one case can be a good artifact in another.
It should just be an improvement for the human player.
Later, when an extended inventory could be implemented (https://github.com/ihhub/fheroes2/discussions/2995), this will no longer be a problem, because all artifacts can be kept, and active or not depending on the case.

> Another issue to solve: why negative artifacts can't be removed? It needs a proper explanation from the game.

Cursed artifacts have to remain cursed.
And so it should not be possible to get rid of them easily, or at least not in this way, they should remain constraining.
Also, I think there is a kind of bug in the original game by the possibility to be able to transfer them to another hero.
That's why I added this option in my mod: https://github.com/ihhub/fheroes2/discussions/3078.

The button should be grayed out when the artifact is a cursed artifact, with a right-click information window indicating why (e.g. "A cursed artifact cannot be discarded.").

> Also we have PoL artifacts which have positive and negative effects. How are we going to handle them?

PoL artifacts which positive and negative effects are considered cursed artifacts (they can be removed when the hero visits the Alchemist's Tower).

These PoL artifacts are considered cursed artifacts:
- Arm of the Martyr
- Broach of Shielding
- Heart of Fire
- Heart of Ice

**Reply by oleg-derevenetz** [reply]
Hi all,

I agree with @Branikolog. There is already lots of "improvements" in the game from the previous team, not all of them are well thought out and break the game mechanics (or even doesn't work properly), so we regularly get rid of certain options.

However, there are also very successful ones - for example, option that shows order of units in the battle (I just keep it turned on all the time) or option that allows HoMM3-like "wait" mechanics in battle (even though it needs some polishing in regard to the UI - current combination of "wait" and "skip" buttons doesn't look too nice).

Therefore, I'd suggest to carefully consider new "improvement" options and all possible outcomes before implementing them.

**Reply by LeHerosInconnu** [reply]
Hello @oleg-derevenetz,

The problem is that one player will like one option and not another and it will be the opposite for another player.

This is the principle of options, the player chooses the options he likes.
But it is true that only interesting/requested options should be offered to the player.

Later, I will refine my opinion for the options already present in fheroes2 that I think will not be used by the players in this thread: https://github.com/ihhub/fheroes2/issues/1272.
